### ISSUE TITLE
Audio "clicking" when in background.

### ISSUE THREAD

**Issue by abdulocracy** [jellyfin/jellyfin-android#134]
On latest beta (9) there is an issue with playing music in the background. Foreground play is flawless, but when app is in the background, or when the screen is shut off, the audio has very frequent lags and clicking noises. Both enabled and disabled battery optimization.

LineageOS 15.1, Android 8.1.0

**Comment by anthonylavado** [comment]
This is @dkanada's old nemesis

**Comment by ma-karai** [comment]
https://github.com/jellyfin/jellyfin-android/issues/147

i did not see this post but yeah, same thing. Glad i am not the only one

**Comment by ma-karai** [comment]
I should have said: damn it I thought I was the only one, since then its definitely on my end.

**Comment by steelwolf** [comment]
>  let me know if you can find a reliable way to reproduce the clicking.
I can reproduce it pretty reliably by playing an mp3 file with the Jellyfin app and then switching to any other app as foreground, turning off the screen, or locking the device.

- Battery optimizations disabled
- Google Pixel 2 XL
- Android 10 (QP1A.190711.020)

**Comment by dkanada** [comment]
Yeah, I'm pretty sure it's based on the device or a certain piece of software, because I haven't seen it at all with the exact same steps you listed. That means I really can't debug the issue so either someone who has the issue has to find the solution or we can wait until someone implements native music playback.

**Comment by shihjay2** [comment]
I can also reproduce this when playing an mp3 with the Jellyfin app and then switching to another app in the foreground, turning off the screen, or locking the device.  I have Nokia 6.1 with Android 9 with the September, 2019 security patches.

**Comment by misterwrong** [comment]
I can also reproduce this when playing flac files with this app, even when the device is not locked. The device is a LG V40 Thinq with Android 9 with the September patches.

**Comment by ma-karai** [comment]
For me this is still happening on 10.4.0_1 I tried a couple of devices all of them show this.

**Comment by anthonylavado** [comment]
>For me this is still happening on 10.4.0_1 I tried a couple of devices all of them show this.

@ma-karai Yeah, this wont be fixed by the server updating, it’s an issue with the app code itself. Essentially, we’re seeing a problem where the system is throttling the web view in the app, so it’s having trouble playing audio.

As @dkanada mentioned, this will likely require native audio playback to be implemented, which we don’t have right now.

**Comment by vincent-peugnet** [comment]
Same problem on my Nokia 6.1 (android One). So I'm using the web interface instead.

**Comment by chillyshackk** [comment]
I have stuttering with any mp3 files I play, wether its 128 or 320kbps, and wether playback is set to auto or 320. On pixel 3 xl.

**Comment by Kabsky** [comment]
hello,
Same audio problem here, audio clicking/stuttering.

- Battery optimizations disabled for jellyfin
- Xiaomi Redmi Note 7
- Android 9 (PKQ1)
- No problem playing the same file from embeded player or Aimp app (emby too).

Btw, thank's for your work on Jellyfin !

**Comment by dkanada** [comment]
This issue is basically frozen until someone wants to implement audio playback using the existing ExoPlayer code @vitorsemeano added. We know this is annoying but it seems to be a general issue with Cordova and the system WebView, so the solution will just be adding native support for audio playback.

**Comment by Z3ROGRAV1TYx** [comment]
Can confirm issue is present on LGv40.

I do notice it seems like it has no issues when running through bluetooth for whatever reason.
Speaker playback seems to have issues, and headphone playback seems to have the issue.

Hope to see this fixed soon. Seems pretty important.

**Comment by bradgy** [comment]
I found my people! Been searching too long for "choppy" audio, not stutter, or clicking :| 

Just upgraded from a Nexus 6P running LineageOS 15 (based on Android 8) to a Moto G7 running Lineage 16 (based on Pie). No problems with music playback before, now once the screen locks or the app goes into the background, it gets very choppy immediately. Both FLAC and mp3 sources. Battery opt. disabled as well. 

Happens in the Android app as well as the web app in Fennec/Firefox.

**Comment by Z3ROGRAV1TYx** [comment]
See this has been here since April, is this being worked on or is it a stagnant issue?
It kinda makes music on Jellyfin pointless tbh

**Comment by ma-karai** [comment]
Unfortunately this is an issue that is difficult to recreate, a few user have it and lots of users don't, ergo its difficult to fix for the jf team :(

**Comment by dkanada** [comment]
It's not a matter of fixing the issue, if someone wanted to implement a native player for music using the ExoPlayer backend it would likely be resolved. That would just require a nontrivial amount of time, which is why no one has done so yet.

**Comment by dbogdanov** [comment]
Same problem for me:
- Battery optimizations disabled for jellyfin
- Pixel 3a
- Android 10

**Comment by rexbron** [comment]
Same issue on Pixel 3a w/ Android 10.

**Comment by henrikh1998** [comment]
Same issue here:
- Oneplus 7 Pro
- Android 10
- Battery Optimization makes no difference
- Audio file format makes no difference (mp3/flac)

**Comment by JettScythe** [comment]
Same issue. 

- Asus Zenfone Max Pro M2

- Android 9 as well as LineageOS 16.0 for MicroG

- Battery optimizations / audio format seem to make make no difference

**Comment by ido1990** [comment]
Same here.
The audio isn't smooth.

- Oneplus 5
- Android 9 (OxygenOS 9.0.11)
- Battery Optimization makes no difference

**Comment by MakisChristou** [comment]
Same here 
- Oneplus 6t
- Android 10 (OxygenOS 10.3.3)
- Battery Optimization makes no difference

**Comment by Wonderfall** [comment]
Same issue :
- Google Pixel 3a
- Android 10 (Pixel Experience, now GrapheneOS)

This is heartbreaking as it pretty much makes Jellyfin unusable on mobile for streaming music, yet I kinda like streaming music from Jellyfin on my desktop.

I would gladly help if I could, but all I can is encouraging you to make Jellyfin even better, and thanking for the great work you've already done! :)

**Comment by DJWhitegold** [comment]
Jellyfin would be perfect if this major issue was fixed. At the moment it can't be used properly as a music server for android, which is such a shame. 

Google pixel 3a, android 10, jellyfin app, battery optimisation disabled makes no difference.

**Comment by grravity** [comment]
> Jellyfin would be perfect if this major issue was fixed. At the moment it can't be used properly as a music server for android, which is such a shame.
> 
> Google pixel 3a, android 10, jellyfin app, battery optimisation disabled makes no difference.

Currently am using Airsonic for music server (Ultrasonic on Android), their webgui could use some work but imo it's a great service it just needs some tlc. Not a huge community around it rn, but just throwing it out there as it could use a bit of love and could be at least a temporary service to use while waiting on this issue.

**Comment by Wonderfall** [comment]
I'm also keeping an eye on [Jellysub](https://github.com/nvllsvm/jellysub/) which acts as Subsonic API interface for Jellyfin, so we can use decent mobile apps for streaming music (the ones compatible with *sonic API) in pair with Jellyfin.

So in the meantime, I agree we have good temporary solutions, though Jellyfin is so close to be a game-changer for me.

**Comment by abdulocracy** [comment]
I've been using [LMS](https://github.com/epoupon/lms) which has seen some great work recently, also has Subsonic api and works alright. Moved from Airsonic for the more modern web interface.

**Comment by dkanada** [comment]
I wrote [Gelli](https://github.com/dkanada/gelli) specifically for music playback through Jellyfin. It currently supports gapless playback, transcoding, and direct playback. It hasn't been released anywhere yet so there are probably bugs, but feel free to try it out. The issue tracking an F-Droid release is [here](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/6755) if you'd like to add a reaction or something.

**Comment by KristupasSavickas** [comment]
@dkanada thanks, that looks great! For anyone's convenience here's a build that I compiled for myself: https://drive.google.com/file/d/17sAn3y5709darQZu63ioSPQY9sPXja5u/view

**Comment by DJWhitegold** [comment]
Wow, thanks for all your suggestions! I'm new to self-hosting media servers and this is very helpful information. I look forward to trying these out. 
PS. Fantastic community spirit on here!

**Comment by DJWhitegold** [comment]
Sweeeeet, thanks for sharing. It's playing audio on my phone in the background without skipping!
On 4 Jun 2020, at 11:36 am, Kristupas Savickas <notifications@github.com<mailto:notifications@github.com>> wrote:

@dkanada<https://github.com/dkanada> thanks, that looks great! For anyone's convenience here's a build that I compiled for myself: https://drive.google.com/file/d/17sAn3y5709darQZu63ioSPQY9sPXja5u/view

**Comment by DJWhitegold** [comment]
> I wrote [Gelli](https://github.com/dkanada/gelli) specifically for music playback through Jellyfin. It's a native app for Android that currently supports gapless playback, transcoding, and direct playback. It hasn't been released anywhere yet so there are probably bugs, but feel free to try it out. The issue tracking an F-Droid release is [here](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/6755) if you'd like to add a reaction or something.

Great work, it's playing audio perfectly on my phone, without skipping!

**Comment by DJWhitegold** [comment]
For anyone wanting to stream music from jellyfin to android without skipping, try Yatse. It's working for me perfectly and a strong alternative to jellyfin's android app. Very surprised as Yatse was originally a xbmc/kodi remote, but they have clearly expanded the functionality of the app. Thought I'd share as I'm very happy that I can keep the superb jellyfin server for music!

**Comment by dkanada** [comment]
I just finished Azure builds so here's an [official release](https://github.com/dkanada/gelli/releases/tag/v1.0.0) of Gelli for anyone who was looking for a download link! There will still probably be minor issues, but the audio clicking should be completely gone from that client.

**Comment by DJWhitegold** [comment]
> I just finished Azure builds so here's an [official release](https://github.com/dkanada/gelli/releases/tag/v1.0.0) of Gelli for anyone who was looking for a download link! There will still probably be minor issues, but the audio clicking should be completely gone from that client.

This is brilliant and it gets better with every release. Thank you so much for developing a solution to streaming audio from the awesome jellyfin to android.

**Comment by yotoprules** [comment]
> For anyone wanting to stream music from jellyfin to android without skipping, try Yatse. Although it's not open source it's working for me perfectly and a strong alternative to jellyfin's android app. Very surprised as Yatse was originally a xbmc/kodi remote, but they have clearly expanded the functionality of the app. Thought I'd share as I'm very happy that I can keep the superb jellyfin server for music!

Unfortunately it always says my login is incorrect even though I am 100% sure they are correct.
The Gelli app made by "dkanada" seems to work perfectly fine although I wish the GUI had some customisability as I'm not a big fan of how it looks out of the box, I wish it told me more information on the music player itself.
But who am I to complain, it works perfectly, songs load in half a second and from my testing I had 0 issues.
Great work :)

**Comment by dkanada** [comment]
@yotoprules do you have any experience with graphic design? I'm also not keen on the interface so I've been looking for possible improvements but design isn't my area of expertise.

**Comment by ido1990** [comment]
> 
> 
> @yotoprules do you have any experience with graphic design? I'm also not keen on the interface so I've been looking for possible improvements but design isn't my area of expertise.

@dkanada 
I do have some experience with graphic design.

**Comment by dkanada** [comment]
As a quick update for the native app, my F-Droid submission has finally been approved! Here is a [link](https://f-droid.org/en/packages/com.dkanada.gramophone) to get the releases straight from their official repository. @ido1990 feel free to open an issue in my repository if you have any ideas.

**Comment by ma-karai** [comment]
gelli is the best

**Comment by grravity** [comment]
> As a quick update for the native app, my F-Droid submission has finally been approved! Here is a link to get the releases straight from their official repository. @ido1990 feel free to open an issue in my repository if you have any ideas.

Glad to hear this!

**Comment by ido1990** [comment]
> 
> 
> As a quick update for the native app, my F-Droid submission has finally been approved! Here is a [link](https://f-droid.org/en/packages/com.dkanada.gramophone) to get the releases straight from their official repository. @ido1990 feel free to open an issue in my repository if you have any ideas.

Congratulations!
Before I open an issue, I would like to know what are the limits and how far would you go with new design.

**Comment by YossiCohen** [comment]
Happens on Samsung Galaxy S8+
Android 9
Both battery options reproduce this issue

**Comment by DJWhitegold** [comment]
> Happens on Samsung Galaxy S8+
> Android 9
> Both battery options reproduce this issue

No indication of this being fixed any time soon, so I highly recommend the Gelli app: https://github.com/dkanada/gelli
